// @applitools/eyes/dist/enums/IosMultiDeviceTarget.js
"use strict";
/* GENERATED FILE */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IosMultiDeviceTargetEnum = void 0;
var IosMultiDeviceTargetEnum;
(function (IosMultiDeviceTargetEnum) {
    IosMultiDeviceTargetEnum["iPhone_SE_1"] = "iPhone SE (1st generation)";
    IosMultiDeviceTargetEnum["iPhone_SE_2"] = "iPhone SE (2nd generation)";
    IosMultiDeviceTargetEnum["iPhone_SE_3"] = "iPhone SE (3rd generation)";
    IosMultiDeviceTargetEnum["iPhone_6"] = "iPhone 6";
    IosMultiDeviceTargetEnum["iPhone_6_Plus"] = "iPhone 6 Plus";
    IosMultiDeviceTargetEnum["iPhone_7"] = "iPhone 7";
    IosMultiDeviceTargetEnum["iPhone_7_Plus"] = "iPhone 7 Plus";
    IosMultiDeviceTargetEnum["iPhone_8"] = "iPhone 8";
    IosMultiDeviceTargetEnum["iPhone_8_Plus"] = "iPhone 8 Plus";
    IosMultiDeviceTargetEnum["iPhone_X"] = "iPhone X";
    IosMultiDeviceTargetEnum["iPhone_XR"] = "iPhone XR";
    IosMultiDeviceTargetEnum["iPhone_Xs"] = "iPhone Xs";
    IosMultiDeviceTargetEnum["iPhone_Xs_Max"] = "iPhone Xs Max";
    IosMultiDeviceTargetEnum["iPhone_11"] = "iPhone 11";
    IosMultiDeviceTargetEnum["iPhone_11_Pro"] = "iPhone 11 Pro";
    IosMultiDeviceTargetEnum["iPhone_11_Pro_Max"] = "iPhone 11 Pro Max";
    IosMultiDeviceTargetEnum["iPhone_12"] = "iPhone 12";
    IosMultiDeviceTargetEnum["iPhone_12_mini"] = "iPhone 12 mini";
    IosMultiDeviceTargetEnum["iPhone_12_Pro"] = "iPhone 12 Pro";
    IosMultiDeviceTargetEnum["iPhone_12_Pro_Max"] = "iPhone 12 Pro Max";
    IosMultiDeviceTargetEnum["iPhone_13"] = "iPhone 13";
    IosMultiDeviceTargetEnum["iPhone_13_mini"] = "iPhone 13 mini";
    IosMultiDeviceTargetEnum["iPhone_13_Pro"] = "iPhone 13 Pro";
    IosMultiDeviceTargetEnum["iPhone_13_Pro_Max"] = "iPhone 13 Pro Max";
    IosMultiDeviceTargetEnum["iPhone_14"] = "iPhone 14";
    IosMultiDeviceTargetEnum["iPhone_14_Pro"] = "iPhone 14 Pro";
    IosMultiDeviceTargetEnum["iPhone_14_Plus"] = "iPhone 14 Plus";
    IosMultiDeviceTargetEnum["iPhone_14_Pro_Max"] = "iPhone 14 Pro Max";
    IosMultiDeviceTargetEnum["iPhone_15"] = "iPhone 15";
    IosMultiDeviceTargetEnum["iPhone_15_Pro"] = "iPhone 15 Pro";
    IosMultiDeviceTargetEnum["iPhone_15_Plus"] = "iPhone 15 Plus";
    IosMultiDeviceTargetEnum["iPhone_15_Pro_Max"] = "iPhone 15 Pro Max";
    IosMultiDeviceTargetEnum["iPhone_16"] = "iPhone 16";
    IosMultiDeviceTargetEnum["iPhone_16_Pro"] = "iPhone 16 Pro";
    IosMultiDeviceTargetEnum["iPhone_16_Plus"] = "iPhone 16 Plus";
    IosMultiDeviceTargetEnum["iPhone_16_Pro_Max"] = "iPhone 16 Pro Max";
})(IosMultiDeviceTargetEnum || (exports.IosMultiDeviceTargetEnum = IosMultiDeviceTargetEnum = {}));
